src/main/java/frontend/DatasetTab.java

```diff
@@ -1,89 +1,104 @@
-        panel.add(fs_panel());
+        panel.add(fileSelectPanel());
         this.add(panel);
+
     }
 
-    public JPanel fs_panel(){
+    private JPanel fileSelectPanel(){
+        //context for the file selector
+        fileSelectContext = new FileContext(this.context, Keys.DatasetFile);
+        WorkflowManager.INSTANCE.registerContext(fileSelectContext);
+
         JPanel panel = new JPanel();
         panel.setLayout(new GridLayout());
-        //JTextPane txt = new JTextPane();
-        JScrollPane jsp = new JScrollPane(txt);
-        JButton fs = new JButton("Browse...");
+        JScrollPane scrollPane = new JScrollPane(txt);
+        JButton fileSelect = new JButton("Browse...");
         JLabel tip = new JLabel("Select a Dataset: ");
-        fs.addActionListener(new fslistener());
+
+        /*
+         * listeners should be anonymous inner classes, this prevents nasty hacks
+         */
+        fileSelect.addActionListener(new ActionListener() {
+            private AbstractCompositeContext context;
+
+            @Override
+            public void actionPerformed(ActionEvent e) {
+                try {
+                    context = WorkflowManager.INSTANCE.getContextByKey(Keys.DatasetFile);
+                    selectFile();
+                    handleFileSelectContext((AbstractParameterContext) context);
+                } catch (IOException exception) {
+                    MiniMLLogger.INSTANCE.exception(exception);
+                }
+            }
+        });
+
         panel.add(tip);
-        panel.add(fs);
-        panel.add(jsp);
-        MiniMLLogger.INSTANCE.info((dataset));
+        panel.add(fileSelect);
+        panel.add(scrollPane);
         return panel;
     }
 
-    public String fs_select() throws IOException{
-        String contents = null;
+    //TODO: move this to wrapper magic class that does not exist yet
+    public FileContext getFileSelectContext(){
+        return this.fileSelectContext;
+    }
+
+    public void handleFileSelectContext(AbstractParameterContext context) {
+        context.setValue(this.dataset, "str");
+        context.updateState();
+    }
+
+    public void selectFile() throws IOException {
+        String contents;
         JFileChooser fileChooser = new JFileChooser();
-        File selectedFile = null;
+        File selectedFile;
         fileChooser.setCurrentDirectory(new File(System.getProperty("user.home")));
         int result = fileChooser.showOpenDialog(this.getParent());
         if (result == JFileChooser.APPROVE_OPTION) {
             selectedFile = fileChooser.getSelectedFile();
-            //System.out.println("Selected file: " + selectedFile.getAbsolutePath());
-            contents = readFileToString(selectedFile);
+            MiniMLLogger.INSTANCE.info("Selected file: " + selectedFile.getAbsolutePath());
+            contents = readFileToString(selectedFile);//TODO: deprecated fileread, why! this will explode on a huge file!
```

Comment:
What are our realistic alternatives? Do we want to handle this is a stream? That seems like a substantially more complex (-> prone to failure) implementation.